### PR TITLE
adding tutorial for ngram training and finetuning

### PR DESCRIPTION
**Pull request by sarane22**
A Tutorial for ngarm training and finetuning with KenLM and evaluating it with NeMo

### CODE REVIEW

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 1]
```diff
@@ -0,0 +1,3133 @@
...
+      "attachments": {},
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BqDDGrWHN4YK"
+      },
+      "source": [
+        "### N-gram Language Model\n",
+        "There are primarily two types of language models:\n",
+        "\n",
+        "- **N-gram language models**: These models use frequency of n-grams to learn the probability distribution over words. Two benefits of N-gram language models are simplicity and scalability – with a larger `n`, a model can store more context with a well-understood space–time tradeoff, enabling small experiments to scale up efficiently.\n",
```
Comment (virajkarandikar) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:36]: can we use `n-gram` or `N-gram` consistently?
may be `n-gram` is more commonly used?

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 2]
```diff
@@ -0,0 +1,3133 @@
...
+        "- **Neural language models**: These models use different kinds of neural networks to model the probability distribution over words, and have surpassed the N-gram language models in the ability to model language, but are generally slower to evaluate.\n",
```
Comment (virajkarandikar) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:37]: What do you mean by `slower to evaluate` ? Did you mean slower inference?

Comment (sarane22) [reply asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:37]: Yes

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb (hunk 1 above)
Comment (sarane22) [reply asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb]: Updated! Replaced `N-gram` with `n-gram`

**Review by virajkarandikar — APPROVED** [review]

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 3]
```diff
@@ -0,0 +1,3135 @@
...
+        "### n-gram Language Model\n",
+        "There are primarily two types of language models:\n",
+        "\n",
+        "- **n-gram language models**: These models use frequency of n-grams to learn the probability distribution over words. Two benefits of n-gram language models are simplicity and scalability – with a larger `n`, a model can store more context with a well-understood space–time tradeoff, enabling small experiments to scale up efficiently.\n",
+        "- **Neural language models**: These models use different kinds of neural networks to model the probability distribution over words, and have surpassed the n-gram language models in the ability to model language, but are generally slower to evaluate.\n",
+        "\n",
+        "In this tutorial, we will show how to train, evaluate, and optionally fine-tune an [n-gram language model](https://web.stanford.edu/~jurafsky/slp3/3.pdf) leveraging TAO Toolkit.\n",
+        "\n",
+        "Prerequisite: You have access and are logged into NVIDIA NGC. For step-by-step instructions, refer to the [NGC Getting Started Guide](https://docs.nvidia.com/ngc/ngc-overview/index.html#registering-activating-ngc-account)."
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:41]: I think this should be its own sub-section, not just a statement.

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 4]
```diff
@@ -0,0 +1,3135 @@
...
+        "from nemo.collections.asr.metrics.wer import CTCDecodingConfig"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "fRcej90lvLxi"
+      },
+      "source": [
+        "### Download English Conformer model from NGC"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:337]:
```suggestion
        "### Download the English Conformer Model from NGC"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 5]
```diff
@@ -0,0 +1,3135 @@
...
+        "!source ~/.bash_profile && ngc registry model download-version \"nvidia/riva/speechtotext_en_us_conformer:trainable_v3.1\" --dest $MODELS_DIR"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "G4w1PlnUvXQm"
+      },
+      "source": [
+        "### Create Lexicon file for flashlight decoder"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:361]:
```suggestion
        "### Create a Lexicon File for Flashlight Decoder"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 6]
```diff
@@ -0,0 +1,3135 @@
...
+        "!python NeMo/scripts/asr_language_modeling/ngram_lm/create_lexicon_from_arpa.py --arpa $RESULTS_DIR/base_lm.arpa --model $MODELS_DIR/speechtotext_en_us_conformer_vtrainable_v3.1/Conformer-CTC-L_spe-1024_en-US_Riva-ASR-SET-3.1.nemo --lower --dst $RESULTS_DIR"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "zlwsoc4ivhsK"
+      },
+      "source": [
+        "### Updating decoder type, Language model and lexicon"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:385]:
```suggestion
        "### Updating the Decoder Type, Language Model, and Lexicon"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 7]
```diff
@@ -0,0 +1,3135 @@
...
+        "transcribe_json(asr_model, f\"{DATA_DOWNLOAD_DIR}/healthcare_eval_set_v1.0/healthcare.json\", f\"{RESULTS_DIR}/healthcare_base_lm.json\")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "1J2w-2gqOyl1"
+      },
+      "source": [
+        "### Calculate word error rate\n"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:505]:
```suggestion
        "### Calculate the Word Error Rate\n"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 8]
```diff
@@ -0,0 +1,3135 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "pZfP4JNyPp4U"
+      },
+      "source": [
+        "---\n",
+        "<a id='isc-Finetuning'></a>\n",
+        "## Finetuning/Interpolation\n"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:580]:
```suggestion
        "## Fine-Tuning and Interpolation\n"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 9]
```diff
@@ -0,0 +1,3135 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "9VARkia8Pstv"
+      },
+      "source": [
+        "The fine-tuning process will continue training using a previously trained model by training a second model on new domain data and interpolating it with the original model. Finetuning requires the original model have intermediate enabled during training. A finetuned model cannot be used for finetuning again. <br>\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:589]:
```suggestion
        "The fine-tuning process will continue training using a previously trained model by training a second model on new domain data and interpolating it with the original model. Fine-tuning requires that the original model have intermediate enabled during training. A fine-tuned model cannot be used for fine-tuning again. <br>\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 10]
```diff
@@ -0,0 +1,3135 @@
...
+        "\n",
+        "\n",
+        "### Downloading and procesing domain data (healthcare) for LM finetuning\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:592]:
```suggestion
        "### Downloading and Processing Domain Data (healthcare) for LM Fine-Tuning\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 11]
```diff
@@ -0,0 +1,3135 @@
...
+        "For the purpose of finetuning on healthcare domain we can make use of Kaggle dataset PubMed 200k RCT: a Dataset for Sequential Sentence Classification in Medical Abstracts. [https://arxiv.org/abs/1710.06071] <br>\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:593]:
```suggestion
        "For the purpose of fine-tuning on a healthcare domain, we can make use of the [Kaggle dataset PubMed 200k RCT: a Dataset for Sequential Sentence Classification in Medical Abstracts](https://arxiv.org/abs/1710.06071). <br>\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 12]
```diff
@@ -0,0 +1,3135 @@
...
+        "This dataset is available at https://www.kaggle.com/datasets/anshulmehtakaggl/200000-abstracts-for-seq-sentence-classification <br>\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:594]:
```suggestion
        "This [dataset is available here](https://www.kaggle.com/datasets/anshulmehtakaggl/200000-abstracts-for-seq-sentence-classification). <br>\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 13]
```diff
@@ -0,0 +1,3135 @@
...
+        "Please follow the instructions to install and authenticate Kaggle API https://www.kaggle.com/docs/api \n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:595]:
```suggestion
        "Perform the instructions to [install and authenticate Kaggle API](https://www.kaggle.com/docs/api). \n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 14]
```diff
@@ -0,0 +1,3135 @@
...
+        "<br>\n",
+        "**Note**: *Each user is responsible for checking the content of datasets and the applicable licenses and determining if suitable for the intended use*\n"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:597]:
```suggestion
        "**Note**: Each user is responsible for checking the content of datasets and the applicable licenses and determining if they are suitable for the intended use.\n"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 15]
```diff
@@ -0,0 +1,3135 @@
...
+      ]
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "9wQQinzHP-jG"
+      },
+      "source": [
+        "The fine-tuning process will continue training using a previously trained model by training a second model on new domain data and interpolating it with the original model. Finetuning requires the original model have intermediate enabled during training. A finetuned model cannot be used for finetuning again. <br>\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:667]:
```suggestion
        "The fine-tuning process will continue training using a previously trained model by training a second model on new domain data and interpolating it with the original model. Fine-tuning requires the original model to have intermediate enabled during training. A fine-tuned model cannot be used for fine-tuning again. <br>\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 16]
```diff
@@ -0,0 +1,3135 @@
...
+        "\n",
+        "\n",
+        "For Finetuning a n-gram language model with KenLM, follow the following steps:\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:670]:
```suggestion
        "For fine-tuning a n-gram language model with KenLM, perform the following steps:\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 17]
```diff
@@ -0,0 +1,3135 @@
...
+        "\n",
+        "\n",
+        "1.   Generate intermediate arpa files.\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:673]:
```suggestion
        "1.   Generate intermediate ARPA files.\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 18]
```diff
@@ -0,0 +1,3135 @@
...
+        "2.   Interpolate\n"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:674]:
```suggestion
        "2.   Interpolate.\n"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb (hunk 17 above)
Comment (LynseyFabel) [reply asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:673]: we should tell them how to do step 1 - link to section for more information

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb (hunk 18 above)
Comment (LynseyFabel) [reply asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:674]: we should make this a full sentence and link to section for more information

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 19]
```diff
@@ -0,0 +1,3135 @@
...
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BAueBCRkWSjY"
+      },
+      "source": [
+        "### Generating internediate arpa"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:683]:
```suggestion
        "### Generating Intermediate ARPA"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 20]
```diff
@@ -0,0 +1,3135 @@
...
+      "metadata": {
+        "id": "fLPjZHq1WjVo"
+      },
+      "source": [
+        "#### Interpolate\n",
+        "Weights for interpolation can be passed with\n",
+        "```\n",
+        " -w 0.6 0.4\n",
+        "```\n",
+        "Here, 60% weightage is assigned to base LM and 40% to Domain"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:718]:
```suggestion
        "Here, 60% weightage is assigned to the base LM and 40% to the domain."
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 21]
```diff
@@ -0,0 +1,3135 @@
...
+        "### Pruning"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "qR69XB7iWuR3"
+      },
+      "source": [
+        "LM generated by simply passing the text corpus to TAO toolkit contains some n-grams which are less frequent(in corpus) and thus have very low probabilities. Such n-grams can be removed by `pruning`.<br>\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:876]:
```suggestion
        "LM generated by simply passing the text corpus to TAO Toolkit contains some n-grams which are less frequent (in corpus) and thus have very low probabilities. Such n-grams can be removed by `pruning`.<br>\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 22]
```diff
@@ -0,0 +1,3135 @@
...
+        "Pruning requires some thresholds which can be passed with `–prune` parameter followed with space separated thresholds specifying count thresholds for each order while generating ARPA:\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:877]:
```suggestion
        "Pruning requires some thresholds which can be passed with the `–prune` parameter followed by space separated thresholds that specify the count thresholds for each order while generating ARPA:\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 23]
```diff
@@ -0,0 +1,3158 @@
...
+        "\n",
+        "In this tutorial, we will show how to train, evaluate, and optionally fine-tune an [n-gram language model](https://web.stanford.edu/~jurafsky/slp3/3.pdf) leveraging NeMo.\n"
+      ]
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "### Prerequisites\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:45]: Add after the section title and before the numbered list. "Ensure you meet the following prerequisites."

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 24]
```diff
@@ -0,0 +1,3158 @@
...
+        "1. You have access and are logged into NVIDIA NGC. For step-by-step instructions, refer to the [NGC Getting Started Guide](https://docs.nvidia.com/ngc/ngc-overview/index.html#registering-activating-ngc-account).\n",
+        "2. You have installed Kaggle API. For step-by-step instructions, refer this [install and authenticate Kaggle API](https://www.kaggle.com/docs/api)."
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:47]:
```suggestion
        "2. You have installed Kaggle API. For step-by-step instructions, refer to this [install and authenticate Kaggle API](https://www.kaggle.com/docs/api)."
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 25]
```diff
@@ -0,0 +1,3158 @@
...
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "7ncrDqn_OLo2"
+      },
+      "source": [
+        "### Installing and Setting up KenLM\n",
+        "Installing KenLM from source"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:99]:
```suggestion
        "Installing KenLM from source."
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 26]
```diff
@@ -0,0 +1,3158 @@
...
+      "attachments": {},
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "RPGEE1d3N5Oo"
+      },
+      "source": [
+        "---\n",
+        "<a id='isc-prepare-data'></a>\n",
+        "### Preparing the Dataset\n",
+        "#### Librispeech LM Normalized Dataset\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:144]:
```suggestion
        "#### LibriSpeech LM Normalized Dataset\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 27]
```diff
@@ -0,0 +1,3158 @@
...
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "_yY-KRYaRFni",
+        "outputId": "4d9ccfc4-e478-4d8e-80d8-50c8e9ff08ef"
+      },
+      "outputs": [],
+      "source": [
+        "# NOTE: Ensure that wget and unzip utilities are available. If not, install them.\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:178]:
```suggestion
        "# Note: Ensure that `wget` and unzip utilities are available. If not, install them.\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 28]
```diff
@@ -0,0 +1,3158 @@
...
+      "attachments": {},
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "7kY0Deemo938"
+      },
+      "source": [
+        "`$KENLM_BASE/build_binary`<br>\n",
+        "Arguments:\n",
+        "\n",
+        "*   `-q`: Quantization flag for the probabilities. (For example `-q 8` store 8 bits of probability)\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:298]:
```suggestion
        "*   `-q`: Quantization flag for the probabilities. For example, `-q 8` store 8 bits of probability.\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 29]
```diff
@@ -0,0 +1,3158 @@
...
+        "*   `-b`: Quantization flag for back-off weights. (For example `-b 7` store 7 bits of back-off)\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:299]:
```suggestion
        "*   `-b`: Quantization flag for back-off weights. For example, `-b 7` store 7 bits of back-off.\n",
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 30]
```diff
@@ -0,0 +1,3158 @@
...
+        "!$KENLM_BASE/build_binary  trie -q 8 -b 7 -a 256 $RESULTS_DIR/base_lm.arpa $RESULTS_DIR/base_lm.bin"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "7KORFpRXuOP1"
+      },
+      "source": [
+        "## Load ASR model"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:325]:
```suggestion
        "## Load the ASR Model"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 31]
```diff
@@ -0,0 +1,3158 @@
...
+      ]
+    },
+    {
+      "attachments": {},
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "zlwsoc4ivhsK"
+      },
+      "source": [
+        "### Updating the Decoder Type, Language Model and Lexicon"
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:402]:
```suggestion
        "### Updating the Decoder Type, Language Model, and Lexicon"
```

File: asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb [hunk 32]
```diff
@@ -0,0 +1,3158 @@
...
+        "id": "9VARkia8Pstv"
+      },
+      "source": [
+        "The fine-tuning process will continue training using a previously trained model by training a second model on new domain data and interpolating it with the original model. Fine-tuning requires that the original model have intermediate enabled during training. A fine-tuned model cannot be used for fine-tuning again. <br>\n",
+        "\n",
+        "\n",
+        "### Downloading and Procesing Domain Data (healthcare) for LM Fine-Tuning\n",
+        "For the purpose of fine-tuning on a healthcare domain we can make use of the [Kaggle dataset PubMed 200k RCT: a Dataset for Sequential Sentence Classification in Medical Abstracts](https://arxiv.org/abs/1710.06071). <br>\n",
+        "This dataset is available [here](https://www.kaggle.com/datasets/anshulmehtakaggl/200000-abstracts-for-seq-sentence-classification). <br>\n",
+        "Perform the instructions to [install and authenticate Kaggle API](https://www.kaggle.com/docs/api).\n",
```
Comment (LynseyFabel) [inline asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb:615]:
```suggestion
        "Follow the instructions to [install and authenticate Kaggle API](https://www.kaggle.com/docs/api).\n",
```

**Comment by sarane22** [comment]
Thanks for the review and approval @LynseyFabel 
Updated the notebook with all the changes.

Thanks.

**Pull request merged**